### ISSUE TITLE
Am I doing this right?

### ISSUE THREAD

**Issue by dejanmarketing** [atlantis-nova/ft-Q#1]
My test show loss of embedding quality when tested on cosine similarities between raw and quantized embeddings. I'm a noob so may have done something wrong.

**Cosine Similarity of Levels 1-5 with Level 0 (Sorted):**

- Similarity between Level 0 and Level 1: 0.9885
- Similarity between Level 0 and Level 2: 0.9546
- Similarity between Level 0 and Level 3: 0.9490
- Similarity between Level 0 and Level 5: 0.8995
- Similarity between Level 0 and Level 4: 0.8443

**Cosine Similarity of Levels 1-5 with Level 0 (Quantized Embeddings, Sorted):**

- Similarity between Level 0 and Level 1: 0.8338
- Similarity between Level 0 and Level 3: 0.6227
- Similarity between Level 0 and Level 2: 0.5890
- Similarity between Level 0 and Level 5: 0.4005
- Similarity between Level 0 and Level 4: 0.2922

Test code:

`import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity

# Load quantized embeddings
quantized_file = "quantized_sentence_embeddings.csv"
quantized_df = pd.read_csv(quantized_file)

# Extract levels (labels) and quantized embeddings
labels = quantized_df["level"].values
quantized_embeddings = quantized_df.iloc[:, 2:].values  # Assuming embeddings start from the 3rd column

# Compute cosine similarity matrix
quantized_similarity_matrix = cosine_similarity(quantized_embeddings)

# Compare levels 1–5 with level 0 and sort by cosine similarity
quantized_results = []
for i in range(1, len(labels)):  # Start from 1 (exclude 0 itself)
    similarity = quantized_similarity_matrix[0, i]
    quantized_results.append((labels[0], labels[i], similarity))

# Sort results by similarity (descending order)
quantized_sorted_results = sorted(quantized_results, key=lambda x: x[2], reverse=True)

# Print sorted results
print("Cosine Similarity of Levels 1-5 with Level 0 (Quantized Embeddings, Sorted):\n")
for base, compare, sim in quantized_sorted_results:
    print(f"Similarity between Level {base} and Level {compare}: {sim:.4f}")
`

Quantizing Code:

`import pandas as pd
import numpy as np
from quantization import *

# Configuration
input_file = "sentence_embeddings.csv"
output_file = "quantized_sentence_embeddings.csv"
interval_mapping_method = "linear"  # Supported: 'linear', 'density'
quantization_method = "binary"     # Supported: 'int8', 'int4', 'int2', 'binary'

# Load the embeddings CSV file
df = pd.read_csv(input_file)

# Separate metadata (level and description) and embeddings
metadata_columns = ["level", "description"]
metadata = df[metadata_columns]
embeddings = df.drop(columns=metadata_columns).values

# Prepare ft-Q quantizer (only needs to run once)
quantizer_list = []
for feature_n in range(embeddings.shape[1]):  # For each embedding dimension
    feature_array = embeddings[:, feature_n].copy()
    scaled_arr, scaled_arr_q = feature_mapping(feature_array, method=interval_mapping_method, data_type=quantization_method)
    quantizer_list.append({'scaled_arr': scaled_arr, 'scaled_arr_q': scaled_arr_q})

# Quantize embeddings
quantized_embeddings = quantize_vector(embeddings, quantizer_list).round()

# Convert quantized embeddings back to a DataFrame
quantized_embeddings_df = pd.DataFrame(quantized_embeddings, columns=[f"embedding_dim_{i}" for i in range(quantized_embeddings.shape[1])])

# Combine metadata and quantized embeddings
output_df = pd.concat([metadata, quantized_embeddings_df], axis=1)

# Save the quantized embeddings to a CSV file
output_df.to_csv(output_file, index=False)

print(f"Quantized embeddings saved to {output_file}")
`

Embed Code:

`import torch
from transformers import AlbertTokenizer, AlbertModel
import pandas as pd

# Load ALBERT model and tokenizer
model_name = "albert-base-v2"
tokenizer = AlbertTokenizer.from_pretrained(model_name)
model = AlbertModel.from_pretrained(model_name)
model.eval()  # Set model to evaluation mode

# Input data: sentences with varying similarity levels
sentences = {
    "level": [0, 1, 2, 3, 4, 5],
    "description": [
        "The cat sat on the mat and stared at the door.",
        "The cat sat on the mat, staring at the door.",
        "A cat rested on the mat, gazing at the door.",
        "The feline lounged on the mat, looking towards the door.",
        "A quiet cat rested nearby, its eyes fixed on something outside.",
        "Across the room, a dog barked at the sound of footsteps on the porch."
    ]
}

# Convert to DataFrame for easier handling
df = pd.DataFrame(sentences)

# Function to generate embeddings
def generate_embedding(sentence):
    inputs = tokenizer(sentence, return_tensors="pt", truncation=True, padding=True)
    with torch.no_grad():  # Disable gradient computation
        outputs = model(**inputs)
        # Use the last hidden state and take the mean over tokens for sentence embedding
        embedding = torch.mean(outputs.last_hidden_state, dim=1)
    return embedding.numpy().flatten()  # Convert to NumPy array

# Generate embeddings for each sentence
embeddings = []
for sentence in df["description"]:
    embedding = generate_embedding(sentence)
    embeddings.append(embedding)

# Convert embeddings to a DataFrame with one column per embedding dimension
embedding_columns = [f"embedding_dim_{i}" for i in range(len(embeddings[0]))]
embeddings_df = pd.DataFrame(embeddings, columns=embedding_columns)

# Combine levels, descriptions, and embeddings
output_df = pd.concat([df, embeddings_df], axis=1)

# Save embeddings to a CSV file
output_file = "sentence_embeddings.csv"
output_df.to_csv(output_file, index=False)

print(f"Embeddings saved to {output_file}")
`

**Comment by dejanmarketing** [comment]
[scripts.zip](https://github.com/user-attachments/files/17965737/scripts.zip)
[embeddings.zip](https://github.com/user-attachments/files/17965739/embeddings.zip)

**Comment by arditobryan** [comment]
Hi (sorry I only answered on medium):

If ft-Q is inferior compared to regular quantization, what I am thinking is:
- How many samples did you encode to extract the distributions?
Perhaps, if the number of samples are too little they might have been misrepresented.
- Are the quantiles cutting out some relevant values?
I am not using minmax, but quantiles. Using minmax is usually a recipe for distaster (an outlier can misrepresent a distributio), but I am also getting rid of them using quantilles (1, 99), maybe they are cutting some important vaues?
- Are feature distributions constant across samples?
The key principle behind ft-Q is also that the feature distribution when presented with different datasets do not change much (ex. if I encode 40.000 steam games or 40.000 poetry text, the distribution of each feature should be very similar, because at the end is still based on human language), does the principle hold with your embedding as well?
It is also possible (is a theory), that by quantizing the embedding the similarity scores need to be looked in a relative way (which means .98 is equivalent to .83 not when compared as absolutes, but because they represetnt the peak. And even if .29 looks very low, in reality is just that the relative minimum of the similarity scores have been shifted.

What I would do to validate is to plot the feature distribution and check if samples are encoded correctly (ex. in the article I spotted one feature that was converted as only 0, which was very bad).

However, if you are comparing regular data f32 with quantized, then similarity loss is expeved (surely can be minimized, but when you cut information there is always a price)

I hope this helps!
